optimade/server/config.py

```diff
@@ -1,260 +1,95 @@
-        )
-        self.mongo_database = config.get(
-            "BACKEND", "MONGO_DATABASE", fallback=self._DEFAULTS("mongo_database")
-        )
-        self.mongo_uri = config.get(
-            "BACKEND", "MONGO_URI", fallback=self._DEFAULTS("mongo_uri")
-        )
-
-        self.page_limit = config.getint(
-            "SERVER", "PAGE_LIMIT", fallback=self._DEFAULTS("page_limit")
-        )
-        self.page_limit_max = config.getint(
-            "SERVER", "PAGE_LIMIT_MAX", fallback=self._DEFAULTS("page_limit_max")
-        )
-        self.default_db = config.get(
-            "SERVER", "DEFAULT_DB", fallback=self._DEFAULTS("default_db")
-        )
-        self.base_url = config.get(
-            "SERVER", "BASE_URL", fallback=self._DEFAULTS("base_url")
-        )
-
-        # This is done in this way, since each field is OPTIONAL
-        self.implementation = {}
-        for field in Implementation.schema()["properties"]:
-            value_config = config.get("IMPLEMENTATION", field, fallback=None)
-            value_default = self._DEFAULTS("implementation")[field]
-            if value_config is not None:
-                self.implementation[field] = value_config
-            elif value_default is not None:
-                self.implementation[field] = value_default
-
-        if "PROVIDER" in config.sections():
-            self.provider = dict(config["PROVIDER"])
-        else:
-            self.provider = self._DEFAULTS("provider")
-
-        self.aliases = {}
-        for endpoint in {"links", "references", "structures"}:
-            if f"{endpoint}.aliases" in config:
-                section = config[f"{endpoint}.aliases"]
-                self.aliases[endpoint] = tuple(section.items())
-
-        self.provider_fields = {}
-        for endpoint in {"links", "references", "structures"}:
-            self.provider_fields[endpoint] = (
-                list({field for field, _ in config[endpoint].items() if _ == ""})
-                if endpoint in config
-                else []
-            )
-
-            # MONGO collections
-            setattr(
-                self,
-                f"{endpoint}_collection",
-                config.get(
-                    "BACKEND",
-                    f"{endpoint.upper()}_COLLECTION",
-                    fallback=self._DEFAULTS(f"{endpoint}_collection"),
-                ),
-            )
+    index_links_path: Path = Path(__file__).parent.joinpath("index_links.json")
 
-    def load_from_json(self):
-        """ Load from the file "config.json", if it exists. """
-        with open(self._path, "r") as f:
-            config = json.load(f)
+    @root_validator(pre=True)
+    def load_default_settings(cls, values):
+        """
+        Loads settings from a root file if available and uses that as defaults in
+        place of built in defaults
+        """
+        config_file_path = Path(values.get("config_file", "~/.optimade.json"))
```

Comment:
Since, it's a `Path` object, it will become absolute when fully invokes, ie when used in an `open` call. I'd rather keep it relative until the very last minute. The idea is that the config_file provides overrides for the default here. It doesn't get written if not present.